File: android-app/app/src/main/java/com/lasohealth/android/core/ml/FeatureEngine.kt
package com.lasohealth.android.core.ml

import com.lasohealth.android.core.analysis.DailySample
import com.lasohealth.android.core.analysis.MetricBaseline
import com.lasohealth.android.core.model.HealthMetric
import java.util.Calendar
import kotlin.math.abs
import kotlin.math.ceil
import kotlin.math.ln
import kotlin.math.log2
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

/**
 * Converts raw health time series into standardised [DailyFeatureVector]s with 18 derived
 * features per metric per day.
 *
 * Uses Welford's online algorithm for incremental z-score normalisation and computes:
 *   raw, rateOfChange, volatility7/14/28, lag1/3/7/14/28/60,
 *   deviationFromBaseline, acceleration, entropy, periodicityStrength,
 *   weekdayOffset, weekendOffset, missingness.
 *
 * Minimum 7 days of data required before any vectors are produced.
 *
 * Ported from iOS `FeatureEngine.swift`.
 */
class FeatureEngine {

    // -- Configuration ------------------------------------------------------

    companion object {
        /** Minimum days of data before producing feature vectors. */
        const val MINIMUM_DAYS = 7

        /** z-score range mapped into entropy bins. */
        private const val ENTROPY_Z_MIN = -3.0
        private const val ENTROPY_Z_MAX = 3.0

        /** Lags evaluated for periodicity strength (autocorrelation). */
        private val PERIODICITY_LAGS = intArrayOf(7, 14, 28)

        /** Lag configurations: (lag days, feature type). */
        private val LAG_CONFIGS: List<Pair<Int, FeatureType>> = listOf(
            1 to FeatureType.LAG_1,
            3 to FeatureType.LAG_3,
            7 to FeatureType.LAG_7,
            14 to FeatureType.LAG_14,
            28 to FeatureType.LAG_28,
            60 to FeatureType.LAG_60,
        )

        /** Sturges' rule with floor 5, cap 15. */
        private fun adaptiveBinCount(sampleSize: Int): Int {
            val sturges = ceil(1.0 + log2(max(sampleSize, 2).toDouble())).toInt()
            return max(5, min(sturges, 15))
        }
    }

    // -- Running statistics (Welford) ---------------------------------------

    /** Incremental mean/variance tracker. */
    private data class WelfordState(
        var count: Int = 0,
        var mean: Double = 0.0,
        var m2: Double = 0.0,
    ) {
        val variance: Double get() = if (count > 1) m2 / count.toDouble() else 0.0
        val stdDev: Double get() = sqrt(variance)

        fun update(value: Double) {
            count++
            val delta = value - mean
            mean += delta / count.toDouble()
            val delta2 = value - mean
            m2 += delta * delta2
        }

        fun zScore(value: Double): Double {
            val sd = stdDev
            return if (sd > 0) (value - mean) / sd else 0.0
        }
    }

    private val runningStats = mutableMapOf<HealthMetric, WelfordState>()
    private val weekdayStats = mutableMapOf<HealthMetric, WelfordState>()
    private val weekendStats = mutableMapOf<HealthMetric, WelfordState>()

    /** Deterministic feature-key ordering produced by the last [extract] call. */
    var orderedKeys: List<FeatureKey> = emptyList()
        private set

    /** Whether the engine has enough data to produce useful features. */
    val isReady: Boolean
        get() = runningStats.values.any { it.count >= MINIMUM_DAYS }

    // -- Public API ---------------------------------------------------------

    /**
     * Build daily feature vectors from raw samples.
     *
     * @param samples  Flat list of daily samples (all metrics mixed).
     * @param baselines  Pre-computed baselines keyed by metric.
     * @return Feature vectors sorted oldest-first, empty if fewer than [MINIMUM_DAYS].
     */
    fun extract(
        samples: List<DailySample>,
        baselines: Map<HealthMetric, MetricBaseline>,
    ): List<DailyFeatureVector> {
        if (samples.isEmpty()) return emptyList()

        val cal = Calendar.getInstance()

        // Group by metric then by date (epoch-millis start-of-day).
        val metricByDate: Map<HealthMetric, Map<Long, Double>> = samples
            .groupBy { it.metric }
            .mapValues { (_, list) ->
                list.associate { startOfDay(it.date, cal) to it.value }
            }

        // Determine date range.
        val allDates = metricByDate.values.flatMap { it.keys }
        val earliest = allDates.minOrNull() ?: return emptyList()
        val today = startOfDay(System.currentTimeMillis(), cal)
        val totalDays = ((today - earliest) / DAY_MS).toInt()
        if (totalDays < MINIMUM_DAYS) return emptyList()

        // Initialise / rebuild running stats.
        for ((metric, dateMap) in metricByDate) {
            val existing = runningStats[metric]
            if (existing == null || existing.count == 0) {
                val state = WelfordState()
                dateMap.values.forEach { state.update(it) }
                runningStats[metric] = state
            }
        }

        // Rebuild weekday/weekend stats from scratch (cheap, ensures correctness).
        weekdayStats.clear()
        weekendStats.clear()
        for ((metric, dateMap) in metricByDate) {
            val wdState = WelfordState()
            val weState = WelfordState()
            for ((date, value) in dateMap) {
                cal.timeInMillis = date
                val dow = cal.get(Calendar.DAY_OF_WEEK)
                if (dow == Calendar.SUNDAY || dow == Calendar.SATURDAY) {
                    weState.update(value)
                } else {
                    wdState.update(value)
                }
            }
            weekdayStats[metric] = wdState
            weekendStats[metric] = weState
        }

        val sortedMetrics = metricByDate.keys.sortedBy { it.name }

        // Build deterministic ordered keys.
        orderedKeys = sortedMetrics.flatMap { metric ->
            FeatureType.entries.map { FeatureKey(metric, it) }
        }

        // Pre-compute z-score series per metric (oldest first) for autocorrelation.
        val zSeriesByMetric = mutableMapOf<HealthMetric, List<Pair<Long, Double>>>()
        for (metric in sortedMetrics) {
            val dateMap = metricByDate[metric] ?: continue
            val stats = runningStats[metric] ?: continue
            if (stats.count < MINIMUM_DAYS) continue
            zSeriesByMetric[metric] = dateMap.entries
                .map { (d, v) -> d to stats.zScore(v) }
                .sortedBy { it.first }
        }

        // Build vectors day-by-day (most recent first, reversed at end).
        val vectors = mutableListOf<DailyFeatureVector>()

        for (offset in 0 until totalDays) {
            val dayMillis = today - offset.toLong() * DAY_MS
            cal.timeInMillis = dayMillis

            val features = mutableMapOf<FeatureKey, Double>()
            var hasAnyData = false

            for (metric in sortedMetrics) {
                val dateMap = metricByDate[metric] ?: continue
                val stats = runningStats[metric] ?: continue
                if (stats.count < MINIMUM_DAYS) continue

                val rawValue = dateMap[dayMillis]

                // Missingness
                if (rawValue == null) {
                    features[FeatureKey(metric, FeatureType.MISSINGNESS)] = 1.0
                    for (ft in FeatureType.entries) {
                        if (ft != FeatureType.MISSINGNESS) {
                            features[FeatureKey(metric, ft)] = FeatureKey.MISSING_SENTINEL
                        }
                    }
                    continue
                }

                hasAnyData = true
                val value = rawValue
                val z = stats.zScore(value)

                features[FeatureKey(metric, FeatureType.MISSINGNESS)] = 0.0
                features[FeatureKey(metric, FeatureType.RAW)] = z

                // Rate of change (day-over-day).
                val prevDate = dayMillis - DAY_MS
                val prevValue = dateMap[prevDate]
                features[FeatureKey(metric, FeatureType.RATE_OF_CHANGE)] =
                    if (prevValue != null) z - stats.zScore(prevValue) else 0.0

                // Acceleration (2nd derivative).
                val prev2Date = dayMillis - 2 * DAY_MS
                val prev1Val = prevValue
                val prev2Val = dateMap[prev2Date]
                if (prev1Val != null && prev2Val != null && stats.count >= 3) {
                    val z1 = stats.zScore(prev1Val)
                    val z2 = stats.zScore(prev2Val)
                    features[FeatureKey(metric, FeatureType.ACCELERATION)] = (z - z1) - (z1 - z2)
                } else {
                    features[FeatureKey(metric, FeatureType.ACCELERATION)] = FeatureKey.MISSING_SENTINEL
                }

                // Rolling volatilities.
                val vol7Vals = collectRecentZScores(dateMap, stats, dayMillis, 7)
                val vol14Vals = collectRecentZScores(dateMap, stats, dayMillis, 14)
                val vol28Vals = collectRecentZScores(dateMap, stats, dayMillis, 28)

                features[FeatureKey(metric, FeatureType.VOLATILITY_7)] =
                    if (vol7Vals.size >= 3) stdDev(vol7Vals) else 0.0

                features[FeatureKey(metric, FeatureType.VOLATILITY_14)] =
                    if (stats.count >= 14 && vol14Vals.size >= 5) stdDev(vol14Vals)
                    else FeatureKey.MISSING_SENTINEL

                features[FeatureKey(metric, FeatureType.VOLATILITY_28)] =
                    if (stats.count >= 28 && vol28Vals.size >= 10) stdDev(vol28Vals)
                    else FeatureKey.MISSING_SENTINEL

                // Lag features.
                for ((lagDays, lagType) in LAG_CONFIGS) {
                    val lagDate = dayMillis - lagDays.toLong() * DAY_MS
                    val lagVal = dateMap[lagDate]
                    features[FeatureKey(metric, lagType)] =
                        if (lagVal != null) stats.zScore(lagVal) else FeatureKey.MISSING_SENTINEL
                }

                // Deviation from baseline.
                val baseline = baselines[metric]
                features[FeatureKey(metric, FeatureType.DEVIATION_BASELINE)] =
                    if (baseline != null && baseline.standardDeviation > 0) {
                        (value - baseline.mean) / baseline.standardDeviation
                    } else {
                        z
                    }

                // Shannon entropy (7-day window).
                features[FeatureKey(metric, FeatureType.ENTROPY)] =
                    if (stats.count >= 7 && vol7Vals.size >= 7) shannonEntropy(vol7Vals)
                    else FeatureKey.MISSING_SENTINEL

                // Periodicity strength.
                val zSeries = zSeriesByMetric[metric]
                if (stats.count >= 14 && zSeries != null && zSeries.size >= 14) {
                    val zVals = zSeries.map { it.second }
                    var maxAbsR = 0.0
                    for (lag in PERIODICITY_LAGS) {
                        if (lag >= zVals.size) continue
                        maxAbsR = max(maxAbsR, abs(autocorrelation(zVals, lag)))
                    }
                    features[FeatureKey(metric, FeatureType.PERIODICITY_STRENGTH)] = maxAbsR
                } else {
                    features[FeatureKey(metric, FeatureType.PERIODICITY_STRENGTH)] = FeatureKey.MISSING_SENTINEL
                }

                // Weekday / weekend offset.
                val wdS = weekdayStats[metric]
                features[FeatureKey(metric, FeatureType.WEEKDAY_OFFSET)] =
                    if (wdS != null && wdS.count >= 3 && stats.stdDev > 0) {
                        (wdS.mean - stats.mean) / stats.stdDev
                    } else {
                        FeatureKey.MISSING_SENTINEL
                    }

                val weS = weekendStats[metric]
                features[FeatureKey(metric, FeatureType.WEEKEND_OFFSET)] =
                    if (weS != null && weS.count >= 2 && stats.stdDev > 0) {
                        (weS.mean - stats.mean) / stats.stdDev
                    } else {
                        FeatureKey.MISSING_SENTINEL
                    }
            }

            if (!hasAnyData) continue

            val context = ContextFeatures.from(dayMillis)
            vectors.add(DailyFeatureVector(date = dayMillis, features = features, context = context))
        }

        // Return sorted oldest-first.
        vectors.sortBy { it.date }
        return vectors
    }

    /**
     * Incremental update with a single new observation (daily update path).
     */
    fun updateIncremental(metric: HealthMetric, newValue: Double) {
        val state = runningStats.getOrPut(metric) { WelfordState() }
        state.update(newValue)
    }

    // -- Private helpers ----------------------------------------------------

    private fun collectRecentZScores(
        dateMap: Map<Long, Double>,
        stats: WelfordState,
        dayMillis: Long,
        windowDays: Int,
    ): List<Double> {
        val result = ArrayList<Double>(windowDays)
        for (d in 0 until windowDays) {
            val pastDate = dayMillis - d.toLong() * DAY_MS
            val v = dateMap[pastDate] ?: continue
            result.add(stats.zScore(v))
        }
        return result
    }

    /**
     * Shannon entropy normalised to [0, 1] using adaptive bin count.
     */
    private fun shannonEntropy(zScores: List<Double>): Double {
        val binCount = adaptiveBinCount(zScores.size)
        val binWidth = (ENTROPY_Z_MAX - ENTROPY_Z_MIN) / binCount
        val bins = IntArray(binCount)
        val n = zScores.size
        if (n == 0) return 0.0

        for (z in zScores) {
            val clamped = z.coerceIn(ENTROPY_Z_MIN, ENTROPY_Z_MAX - 1e-9)
            val idx = ((clamped - ENTROPY_Z_MIN) / binWidth).toInt().coerceIn(0, binCount - 1)
            bins[idx]++
        }

        var entropy = 0.0
        val nDouble = n.toDouble()
        for (count in bins) {
            if (count == 0) continue
            val p = count.toDouble() / nDouble
            entropy -= p * ln(p)
        }
        val maxEntropy = ln(binCount.toDouble())
        return if (maxEntropy > 0) entropy / maxEntropy else 0.0
    }

    /**
     * Autocorrelation of [values] at the given [lag].
     */
    private fun autocorrelation(values: List<Double>, lag: Int): Double {
        val n = values.size
        if (lag >= n || n < 2) return 0.0

        val mean = values.average()
        var numerator = 0.0
        var denominator = 0.0

        for (i in values.indices) {
            val diff = values[i] - mean
            denominator += diff * diff
            if (i + lag < n) {
                numerator += diff * (values[i + lag] - mean)
            }
        }
        return if (denominator > 0) numerator / denominator else 0.0
    }

    private fun stdDev(values: List<Double>): Double {
        if (values.size < 2) return 0.0
        val mean = values.average()
        val variance = values.sumOf { (it - mean) * (it - mean) } / values.size
        return sqrt(variance)
    }

    private fun startOfDay(epochMillis: Long, cal: Calendar): Long {
        cal.timeInMillis = epochMillis
        cal.set(Calendar.HOUR_OF_DAY, 0)
        cal.set(Calendar.MINUTE, 0)
        cal.set(Calendar.SECOND, 0)
        cal.set(Calendar.MILLISECOND, 0)
        return cal.timeInMillis
    }
}

private const val DAY_MS: Long = 86_400_000L

File: Core/Analysis/ML/FeatureEngine.swift
import Foundation

/// Converts raw health time series into standardized daily feature vectors with derived features.
/// Uses Welford's online algorithm for incremental normalization updates.
///
/// Computes per-metric features:
///   raw, roc, vol7, lag1, lag3, lag7, devBaseline (original)
///   acceleration, entropy, periodicityStr, vol14, vol28, lag14, lag28, lag60,
///   weekdayOffset, weekendOffset, missingness (extended)
final class FeatureEngine {
    /// Minimum days of data required before producing feature vectors
    static let minimumDays = 7

    /// Running statistics per metric for incremental z-score normalization
    private var runningStats: [HealthMetric: WelfordState] = [:]

    /// Running weekday/weekend means per metric for offset features (maintained incrementally)
    private var weekdayStats: [HealthMetric: WelfordState] = [:]
    private var weekendStats: [HealthMetric: WelfordState] = [:]
    /// Tracks which dates have been incorporated into weekday/weekend stats (prevents double-counting)
    private var weekdayWeekendProcessedDates: [HealthMetric: Set<Date>] = [:]

    /// Ordered feature keys for deterministic array output
    private(set) var orderedKeys: [FeatureKey] = []

    /// Cached feature vectors from previous builds. Only the most recent `rebuildWindowDays`
    /// are recomputed each call; older days reuse cached vectors.
    private var cachedVectors: [Date: DailyFeatureVector] = [:]
    /// Days within this window are always recomputed (handles late-arriving data).
    private static let rebuildWindowDays = 3

    // MARK: - History Horizon

    /// Newest days that get a feature vector. A first sync can hand back a decade of HealthKit
    /// samples, and building every one of those days costs minutes of CPU, which trips the
    /// pipeline's own thermal guard before any insight is produced. Two years still leaves far
    /// more training rows than any downstream component asks for (the largest need is 14 days).
    private static let maxHistoryDays = 730

    /// Extra older days kept in the day lookup array beyond the build horizon, so the oldest
    /// built day can still read its lag features. The deepest lookup into that array is lag60.
    private static let maxLookbackDays = 60

    // MARK: - Shannon Entropy Constants

    /// Adaptive bin count: uses Sturges' rule k = ceil(1 + log2(n)) with floor of 5
    /// for small samples, rather than a fixed 5 bins regardless of data volume
    private static func adaptiveBinCount(sampleSize: Int) -> Int {
        let sturges = Int(ceil(1.0 + log2(Double(max(sampleSize, 2)))))
        return max(5, min(sturges, 15))
    }
    /// Range of z-scores mapped into bins: [-3, 3].
    private static let entropyZMin = -3.0
    private static let entropyZMax = 3.0

    // MARK: - Periodicity Lags

    /// Lags (in days) at which autocorrelation is evaluated for periodicity strength.
    private static let periodicityLags = [7, 14, 28]
    /// Observations in the trailing window used for periodicity. Long enough to cover three
    /// cycles at the longest lag (28 days) without reaching deep into the past.
    private static let periodicityWindowSamples = 90
    /// Observations up to and including the day being built before periodicity is reported.
    private static let periodicityMinSamples = 14

    // MARK: - Weekday / Weekend Offsets

    /// Weekday observations required before the weekday offset is reported.
    private static let minWeekdaySamples = 3
    /// Weekend observations required before the weekend offset is reported.
    private static let minWeekendSamples = 2

    // MARK: - Hoisted Per-Day Constants

    /// Every feature type except missingness. Hoisted because the per-day loop would otherwise
    /// rebuild `FeatureType.allCases` for every metric on every day.
    private static let nonMissingnessTypes = FeatureType.allCases.filter { $0 != .missingness }

    /// Lag features and the day offset each one reads.
    private static let lagConfigs: [(days: Int, type: FeatureType)] = [
        (1, .lag1), (3, .lag3), (7, .lag7),
        (14, .lag14), (28, .lag28), (60, .lag60),
    ]

    // MARK: - Welford's Online Statistics

    /// Incremental mean/variance tracker
    struct WelfordState: Codable {
        var count: Int = 0
        var mean: Double = 0
        var m2: Double = 0

        var variance: Double {
            count > 1 ? m2 / Double(count) : 0
        }

        var stdDev: Double {
            variance.squareRoot()
        }

        mutating func update(value: Double) {
            count += 1
            let delta = value - mean
            mean += delta / Double(count)
            let delta2 = value - mean
            m2 += delta * delta2
        }

        func zScore(for value: Double) -> Double {
            let sd = stdDev
            guard sd > 0 else { return 0 }
            return (value - mean) / sd
        }
    }

    // MARK: - Feature Extraction

    /// Build daily feature vectors from raw time series data.
    /// Returns vectors sorted by date (oldest first).
    func buildFeatureVectors(
        timeSeries: [HealthMetric: MetricTimeSeries],
        baselines: [HealthMetric: UserBaseline]
    ) -> [DailyFeatureVector] {
        let calendar = Date.cal
        let today = calendar.startOfDay(for: Date())

        // Determine date range from available data
        let allDates = timeSeries.values.flatMap { $0.sortedSamples.map { calendar.startOfDay(for: $0.date) } }
        guard let earliest = allDates.min() else { return [] }

        let totalDays = calendar.dateComponents([.day], from: earliest, to: today).day ?? 0
        guard totalDays >= Self.minimumDays else { return [] }

        // Build date-indexed lookup per metric
        var metricByDate: [HealthMetric: [Date: Double]] = [:]
        for (metric, series) in timeSeries {
            var dateMap: [Date: Double] = [:]
            for sample in series.sortedSamples {
                let day = calendar.startOfDay(for: sample.date)
                dateMap[day] = sample.value
            }
            metricByDate[metric] = dateMap
        }

        let metricsWithData = Array(metricByDate.keys)

        // Update running statistics incrementally. only recompute if no prior state exists
        for (metric, dateMap) in metricByDate {
            if runningStats[metric] == nil || runningStats[metric]!.count == 0 {
                // First time: build from scratch
                var state = WelfordState()
                for value in dateMap.values {
                    state.update(value: value)
                }
                runningStats[metric] = state
            }
            // Otherwise keep existing Welford state. incremental updates happen via updateIncremental()
        }

        // Build weekday/weekend running means per metric incrementally.
        // Only process dates not yet incorporated to avoid O(N) full rebuild each run.
        for (metric, dateMap) in metricByDate {
            if weekdayStats[metric] == nil { weekdayStats[metric] = WelfordState() }
            if weekendStats[metric] == nil { weekendStats[metric] = WelfordState() }
            var processed = weekdayWeekendProcessedDates[metric] ?? []
            for (date, value) in dateMap {
                guard !processed.contains(date) else { continue }
                processed.insert(date)
                let dow = calendar.component(.weekday, from: date) // 1=Sun, 7=Sat
                if dow == 1 || dow == 7 {
                    weekendStats[metric]?.update(value: value)
                } else {
                    weekdayStats[metric]?.update(value: value)
                }
            }
            weekdayWeekendProcessedDates[metric] = processed
        }

        // Build ordered keys for deterministic array output
        let sortedMetrics = metricsWithData.sorted { $0.rawValue < $1.rawValue }
        orderedKeys = sortedMetrics.flatMap { metric in
            FeatureType.allCases.map { FeatureKey(metric: metric, type: $0) }
        }

        // Pre-compute the z-score series per metric (oldest first) plus a date -> index lookup,
        // so each day can take a trailing autocorrelation window that ends on that day.
        var zValuesByMetric: [HealthMetric: [Double]] = [:]
        var zDateIndex: [HealthMetric: [Date: Int]] = [:]
        for metric in sortedMetrics {
            guard let dateMap = metricByDate[metric],
                  let stats = runningStats[metric],
                  stats.count >= Self.minimumDays else { continue }
            let sortedDates = dateMap.keys.sorted()
            var values: [Double] = []
            values.reserveCapacity(sortedDates.count)
            var index: [Date: Int] = [:]
            index.reserveCapacity(sortedDates.count)
            for (i, date) in sortedDates.enumerated() {
                values.append(stats.zScore(for: dateMap[date]!))
                index[date] = i
            }
            zValuesByMetric[metric] = values
            zDateIndex[metric] = index
        }

        // Weekday and weekend offsets depend only on the running stats, never on the day being
        // built, so they are computed once per metric instead of once per metric per day.
        var weekdayOffsetByMetric: [HealthMetric: Double] = [:]
        var weekendOffsetByMetric: [HealthMetric: Double] = [:]
        for metric in sortedMetrics {
            guard let stats = runningStats[metric], stats.stdDev > 0 else { continue }
            if let wdStats = weekdayStats[metric], wdStats.count >= Self.minWeekdaySamples {
                weekdayOffsetByMetric[metric] = (wdStats.mean - stats.mean) / stats.stdDev
            }
            if let weStats = weekendStats[metric], weStats.count >= Self.minWeekendSamples {
                weekendOffsetByMetric[metric] = (weStats.mean - stats.mean) / stats.stdDev
            }
        }

        // Pre-build the date array ONCE to keep Calendar.date(byAdding:) out of the inner loops.
        // allDays[0] = today, allDays[1] = yesterday, and so on. It runs one lookback window past
        // the build horizon so the oldest built day can still read its lag features.
        let lookupDays = min(totalDays, Self.maxHistoryDays + Self.maxLookbackDays)
        var allDays: [Date] = []
        allDays.reserveCapacity(lookupDays)
        for offset in 0..<lookupDays {
            guard let d = calendar.date(byAdding: .day, value: -offset, to: today) else { continue }
            allDays.append(calendar.startOfDay(for: d))
        }

        // Build vectors for each day inside the horizon (newest first)
        var vectors: [DailyFeatureVector] = []
        let buildDays = min(allDays.count, Self.maxHistoryDays)

        for dayOffset in 0..<buildDays {
            let day = allDays[dayOffset]

            // Reuse cached vector for days outside the rebuild window.
            // Only the most recent days need recomputation (new/late data).
            if dayOffset >= Self.rebuildWindowDays,
               let cached = cachedVectors[day] {
                vectors.append(cached)
                continue
            }

            var features: [FeatureKey: Double] = [:]
            var hasAnyData = false

            for metric in sortedMetrics {
                guard let dateMap = metricByDate[metric],
                      let stats = runningStats[metric],
                      stats.count >= Self.minimumDays else { continue }

                let rawValue = dateMap[day]

                // -- Missingness feature (always computable) --
                if rawValue == nil {
                    features[FeatureKey(metric: metric, type: .missingness)] = 1.0
                    // Mark all other features as missing for this metric on this day
                    for ft in Self.nonMissingnessTypes {
                        features[FeatureKey(metric: metric, type: ft)] = FeatureKey.missingSentinel
                    }
                    continue
                }

                let value = rawValue!
                hasAnyData = true
                let zScore = stats.zScore(for: value)

                features[FeatureKey(metric: metric, type: .missingness)] = 0.0

                // --- Raw z-score ---
                features[FeatureKey(metric: metric, type: .raw)] = zScore

                // --- Rate of change (day-over-day) --- uses allDays index instead of Calendar
                let prevRoc: Double
                let prev1Idx = dayOffset + 1
                if prev1Idx < allDays.count,
                   let prevValue = dateMap[allDays[prev1Idx]] {
                    let prevZ = stats.zScore(for: prevValue)
                    prevRoc = zScore - prevZ
                } else {
                    prevRoc = 0
                }
                features[FeatureKey(metric: metric, type: .roc)] = prevRoc

                // --- Acceleration (2nd derivative) --- requires 3+ days
                let prev2Idx = dayOffset + 2
                if stats.count >= 3,
                   prev1Idx < allDays.count,
                   prev2Idx < allDays.count,
                   let prev1Value = dateMap[allDays[prev1Idx]],
                   let prev2Value = dateMap[allDays[prev2Idx]] {
                    let prev1Z = stats.zScore(for: prev1Value)
                    let prev2Z = stats.zScore(for: prev2Value)
                    let roc1 = zScore - prev1Z
                    let roc0 = prev1Z - prev2Z
                    features[FeatureKey(metric: metric, type: .acceleration)] = roc1 - roc0
                } else {
                    features[FeatureKey(metric: metric, type: .acceleration)] = FeatureKey.missingSentinel
                }

                // --- Rolling volatility helpers ---
                // Collect z-score windows using pre-built allDays array (no Calendar calls)
                let vol7Values = collectRecentZScoresIndexed(
                    dateMap: dateMap, stats: stats, dayOffset: dayOffset, windowDays: 7, allDays: allDays
                )
                let vol14Values = collectRecentZScoresIndexed(
                    dateMap: dateMap, stats: stats, dayOffset: dayOffset, windowDays: 14, allDays: allDays
                )
                let vol28Values = collectRecentZScoresIndexed(
                    dateMap: dateMap, stats: stats, dayOffset: dayOffset, windowDays: 28, allDays: allDays
                )

                // --- 7-day rolling volatility ---
                if vol7Values.count >= 3 {
                    let (_, vol) = AccelerateML.welfordVariance(vol7Values)
                    features[FeatureKey(metric: metric, type: .vol7)] = vol.squareRoot()
                } else {
                    features[FeatureKey(metric: metric, type: .vol7)] = 0
                }

                // --- 14-day rolling volatility --- needs 14+ days in running stats
                if stats.count >= 14, vol14Values.count >= 5 {
                    let (_, vol) = AccelerateML.welfordVariance(vol14Values)
                    features[FeatureKey(metric: metric, type: .vol14)] = vol.squareRoot()
                } else {
                    features[FeatureKey(metric: metric, type: .vol14)] = FeatureKey.missingSentinel
                }

                // --- 28-day rolling volatility --- needs 28+ days in running stats
                if stats.count >= 28, vol28Values.count >= 10 {
                    let (_, vol) = AccelerateML.welfordVariance(vol28Values)
                    features[FeatureKey(metric: metric, type: .vol28)] = vol.squareRoot()
                } else {
                    features[FeatureKey(metric: metric, type: .vol28)] = FeatureKey.missingSentinel
                }

                // --- Lag features --- uses allDays index instead of Calendar
                for (lagDays, lagType) in Self.lagConfigs {
                    let lagIdx = dayOffset + lagDays
                    if lagIdx < allDays.count,
                       let lagValue = dateMap[allDays[lagIdx]] {
                        features[FeatureKey(metric: metric, type: lagType)] = stats.zScore(for: lagValue)
                    } else {
                        features[FeatureKey(metric: metric, type: lagType)] = FeatureKey.missingSentinel
                    }
                }

                // --- Deviation from baseline ---
                if let baseline = baselines[metric], baseline.standardDeviation > 0 {
                    features[FeatureKey(metric: metric, type: .devBaseline)] =
                        (value - baseline.mean) / baseline.standardDeviation
                } else {
                    features[FeatureKey(metric: metric, type: .devBaseline)] = zScore
                }

                // --- Shannon entropy of last 7 z-score values (binned into 5 bins) ---
                if stats.count >= 7, vol7Values.count >= 7 {
                    features[FeatureKey(metric: metric, type: .entropy)] =
                        Self.shannonEntropy(zScores: vol7Values)
                } else {
                    features[FeatureKey(metric: metric, type: .entropy)] = FeatureKey.missingSentinel
                }

                // --- Periodicity strength (max |autocorrelation| over a trailing window) ---
                // The window has to end on the day being built. Running it over the whole series
                // folds later days into a past day's feature, which biases every model that
                // trains on these vectors, and it also made the value identical on every day.
                if let zValues = zValuesByMetric[metric],
                   let endIndex = zDateIndex[metric]?[day],
                   endIndex + 1 >= Self.periodicityMinSamples {
                    let startIndex = max(0, endIndex + 1 - Self.periodicityWindowSamples)
                    let window = Array(zValues[startIndex...endIndex])
                    var maxAbsR = 0.0
                    for lag in Self.periodicityLags where lag < window.count {
                        let absR = abs(AccelerateML.autocorrelation(window, lag: lag))
                        if absR > maxAbsR { maxAbsR = absR }
                    }
                    features[FeatureKey(metric: metric, type: .periodicityStr)] = maxAbsR
                } else {
                    features[FeatureKey(metric: metric, type: .periodicityStr)] = FeatureKey.missingSentinel
                }

                // --- Weekday / weekend offsets (precomputed once per metric above) ---
                features[FeatureKey(metric: metric, type: .weekdayOffset)] =
                    weekdayOffsetByMetric[metric] ?? FeatureKey.missingSentinel
                features[FeatureKey(metric: metric, type: .weekendOffset)] =
                    weekendOffsetByMetric[metric] ?? FeatureKey.missingSentinel
            }

            guard hasAnyData else { continue }

            let context = ContextFeatures.from(date: day)
            vectors.append(DailyFeatureVector(date: day, features: features, context: context))
        }

        // Update cache with freshly computed vectors
        for vector in vectors {
            cachedVectors[vector.date] = vector
        }

        return vectors.sorted { $0.date < $1.date }
    }

    // MARK: - Incremental Updates

    /// Update running statistics incrementally with new data (daily update)
    func updateIncremental(metric: HealthMetric, newValue: Double) {
        if runningStats[metric] == nil {
            runningStats[metric] = WelfordState()
        }
        runningStats[metric]?.update(value: newValue)
    }

    // MARK: - Private Helpers

    /// Collect z-scored values using pre-built date array (zero Calendar calls).
    /// `dayOffset` is the index into `allDays` for the current day.
    private func collectRecentZScoresIndexed(
        dateMap: [Date: Double],
        stats: WelfordState,
        dayOffset: Int,
        windowDays: Int,
        allDays: [Date]
    ) -> [Double] {
        var values: [Double] = []
        values.reserveCapacity(windowDays)
        for d in 0..<windowDays {
            let idx = dayOffset + d
            guard idx < allDays.count else { break }
            if let v = dateMap[allDays[idx]] {
                values.append(stats.zScore(for: v))
            }
        }
        return values
    }

    /// Compute Shannon entropy from z-score values binned adaptively over [-3, 3].
    /// Bin count follows Sturges' rule: k = ceil(1 + log2(n)), clamped to [5, 15].
    /// Formula: -Sum(p_i * log(p_i)) for non-zero bins.
    /// Returns entropy in nats, normalized by max entropy = ln(binCount) to [0, 1].
    private static func shannonEntropy(zScores: [Double]) -> Double {
        let binCount = adaptiveBinCount(sampleSize: zScores.count)
        let zMin = entropyZMin
        let zMax = entropyZMax
        let binWidth = (zMax - zMin) / Double(binCount)

        var bins = [Int](repeating: 0, count: binCount)
        let n = zScores.count
        guard n > 0 else { return 0 }

        for z in zScores {
            // Clamp to [zMin, zMax)
            let clamped = Swift.min(Swift.max(z, zMin), zMax - 1e-9)
            let binIndex = Int((clamped - zMin) / binWidth)
            let safeBin = Swift.min(Swift.max(binIndex, 0), binCount - 1)
            bins[safeBin] += 1
        }

        var entropy = 0.0
        let nDouble = Double(n)
        for count in bins {
            guard count > 0 else { continue }
            let p = Double(count) / nDouble
            entropy -= p * log(p)
        }
        // Normalize by max entropy (uniform distribution) so output is scale-invariant [0, 1]
        let maxEntropy = log(Double(binCount))
        return maxEntropy > 0 ? entropy / maxEntropy : 0
    }

}
